src/compiler/checker.ts

```diff
@@ -7596,6 +7596,9 @@ namespace ts {
                 }
                 return anyType;
             }
+            if (containsThisProperty(expression.left, expression.right)) {
+                return anyType;

```

Comment:
Rather than returning `any` on all binary expressions with both lhs and rhs references to the same thing, why not elide such binary expressions from the set of expressions whose types are union's together in the first place? This way the type is still calculated based on all other usages, and not quietly `any`'d?

Reply:
This is a per-declaration type. The caller, `getWidenedTypeFromAssignmentDeclaration`, does something similar to what you suggest, except that it divides declarations into constructor vs non-constructor, then unions all types from the constructor if any exist, otherwise unioning all non-constructor types. See my previous reply to @andrewbranch's question.

I think this is better than dropping circular types because of Andrew's `this.x = 0; this.x = this.x.toString()` example. Ideally, we would give `x: string`, but I think `x: any` is better than `x: number`.

Reply:
Right, but, like, if we never request the type of the declaration, it'll never get flagged as circular. Something like
```ts

                if (!jsdocType) {
                    (types || (types = [])).push(((isBinaryExpression(expression) && !containsSameNamedThisProperty(expression.left, expression.right)) || isCallExpression(expression)) ? getInitializerTypeFromAssignmentDeclaration(symbol, resolvedSymbol, expression, kind) : neverType);
                }
```

Reply:
> I think this is better than dropping circular types because of Andrew's this.x = 0; this.x = this.x.toString() example. Ideally, we would give x: string, but I think x: any is better than x: number.

Moreover, we support
```ts
var x;
x = 1;
x = x.toString();
x;
```
without error, so why not
```ts
this.x = 1;
this.x = this.x.toString();
this.x;
```
if we find that case of value? We'd just need to assign the declared type to the `autoType` when we see a self-referential assignment like this.
```

Reply:
1. Moving containsSameNamedThisProperty earlier is equivalent to returning `never` insteading of `any`. When you do this, though, you get this:

```js
    /** @class */
    function C() {
        this.x = 0;
        this.x = this.x.toString();
        ~~~~~~
!!! error TS2322: Type 'string' is not assignable to type 'number'.
    }
```

Which I don't think is good. I'd rather give `x: any` than `x: number`.

2. I don't understand how you'd make this work with control flow's `autoType`, since I don't think it will work consistently functions. Can you explain?

Reply:
`autoType` is an any-like that's essentially narrowable by assignment in control flow. Just return it here instead of `anyType` and check out the behavior.

Reply:
I'm sceptical that it'll be helpful, but I'm biased by imagining a lot of short methods that would still end up with any. I filed https://github.com/microsoft/TypeScript/issues/37900 to track this, since when I tried just `return autoType`, control flow didn't kick in and it's not obvious why.